client/lib/analytics/ad-tracking.js

```diff
@@ -1363,6 +1417,149 @@ function recordViewCheckoutInCriteo( cart ) {
 	} );
 }
 
+/**
+ * Records that a user viewed the checkout page
+ *
+ * @param {Object} cart - cart as `CartValue` object
+ * @returns {void}
+ */
+function recordViewCheckoutInDonutsGtag( cart ) {
+	if ( ! isAdTrackingAllowed() || ! isDonutsGtagEnabled ) {
+		return;
+	}
+
+	if ( cart.is_free_signup ) {
+		return;
+	}
+
+	recordParamsInDonutsGtag(
+		'conversion',
+		'DC-8907854/cartd0/wpress+unique',
+		cartToDonutsOrderSummary( cart )
+	);
+}
+
+function domainNameToTld( domainName ) {
+	if ( domainName.indexOf( '.' ) === -1 ) {
+		return null;
+	}
+	return domainName.slice( domainName.indexOf( '.' ) + 1 ).toLowerCase();
+}
+
+/**
+ * Converts the products in a cart to the format Donuts expects for its `u90` order summary property
+ *
+ * @param {Object} cart - cart as `CartValue` object
+ * @returns {Array} - An array of items to include in the Criteo tracking call
+ */
+function cartToDonutsOrderSummary( cart ) {
+	const DONUTS_TLDS = [
+		'associates',
+		'business',
+		'careers',
+		'center',
+		'company',
+		'consulting',
+		'enterprises',
+		'gmbh',
+		'group',
+		'industries',
+		'international',
+		'limited',
+		'ltd',
+		'management',
+		'network',
+		'partners',
+		'sarl',
+		'services',
+		'solutions',
+		'support',
+		'ventures',
+		'actor',
+		'band',
+		'bingo',
+		'casino',
+		'dance',
+		'dating',
+		'dog',
+		'events',
+		'games',
+		'live',
+		'media',
+		'movie',
+		'rocks',
+		'show',
+		'singles',
+		'studio',
+		'theater',
+		'video',
+		'camera',
+		'equipment',
+		'gallery',
+		'graphics',
+		'media',
+		'photography',
+		'photos',
+		'pictures',
+		'studio',
+		'video',
+		'agency',
+		'consulting',
+		'digital',
+		'direct',
+		'email',
+		'marketing',
+		'media',
+		'productions',
+		'social',
+		'studio',
+		'video',
+		'chat',
+		'directory',
+		'fyi',
+		'media',
+		'memorial',
+		'news',
+		'reviews',
+		'social',
+		'tips',
+		'video',
+		'zone',
+	];
+	// for each domain: if domain ending is in list of donuts domain endings, add to u99 string
+	// types of products covered: domain registration, whois_privacy
+	// TODO: add support for 'renewal' and 'transfer' products
+	debug( 'cartToDonutsOrderSummary:cart', cart );
+	const domain_registrations = cart.products
+		.filter( p => p.is_domain_registration || p.product_slug === 'private_whois' )
+		.map( p => {
+			let donuts_type = 'unknown';
+			if ( p.is_domain_registration ) {
+				donuts_type = 'registration';
+			} else if ( p.product_slug === 'private_whois' ) {
+				donuts_type = 'whois_privacy';
+			}
+			return Object.assign( p, { tld: domainNameToTld( p.meta ), donuts_type: donuts_type } );
```

Comment:
This is mutating `p`… is that intended? It looks like any `cart.products` that pass the filter will have these properties added.

Should this be creating a new object?

```suggestion
			return Object.assign( {}, p, { tld: domainNameToTld( p.meta ), donuts_type: donuts_type } );
```